packages/opentelemetry-cloud-trace-propagator/src/CloudPropagator.ts

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  HttpTextPropagator,
+  Context,
+  SetterFunction,
+  GetterFunction,
+  TraceFlags,
+} from '@opentelemetry/api';
+import { decToHex, hexToDec } from 'hex2dec';
+import {
+  setExtractedSpanContext,
+  getParentSpanContext,
+} from '@opentelemetry/core';
+
+/**
+ * This file implements propagation for the Stackdriver Trace v1 Trace Context
+ * format.
+ *
+ * The header specification is:
+ * "X-Cloud-Trace-Context: TRACE_ID/SPAN_ID;o=TRACE_TRUE"
+ * Where:
+ *  {TRACE_ID} is a 32-character hexadecimal value representing a 128-bit
+ *    number. It should be unique between your requests, unless you
+ *    intentionally want to bundle the requests together.
+ *  {SPAN_ID} is the decimal representation of the (unsigned) span ID.
+ *    It should be randomly generated and unique in your trace. For subsequent requests,
+ *    set SPAN_ID to the span ID of the parent request.
+ *  {TRACE_TRUE} must be 1 to trace request. Specify 0 to not trace the request.
+ */
+
+/** Header that carries span context across Google infrastructure. */
+export const X_CLOUD_TRACE_HEADER = 'x-cloud-trace-context';
+
+export class CloudPropagator implements HttpTextPropagator {
+  inject(context: Context, carrier: unknown, setter: SetterFunction): void {
+    const spanContext = getParentSpanContext(context);
+    if (!spanContext) return;
+
+    const header = `${spanContext.traceId}/${hexToDec(
+      spanContext.spanId
+    )};o=${spanContext.traceFlags & TraceFlags.SAMPLED}`;
+
+    setter(carrier, X_CLOUD_TRACE_HEADER, header);
+  }
+
+  extract(context: Context, carrier: unknown, getter: GetterFunction): Context {
+    const traceContextHeader = getter(carrier, X_CLOUD_TRACE_HEADER);
+    const traceContextHeaderValue = Array.isArray(traceContextHeader)
+      ? traceContextHeader[0]
+      : traceContextHeader;
+    if (typeof traceContextHeaderValue !== 'string') {
+      return context;
+    }
+    const matches = traceContextHeaderValue.match(
+      /^([0-9a-fA-F]{32})(?:\/([0-9]+))(?:;o=(.*))?/
+    );
+
+    if (!matches) {
+      return context;
+    }
+
+    const spanContext = {
+      traceId: matches[1],
+      spanId: decToHex(matches[2], { prefix: false }).padStart(16, '0'),
+      traceFlags: matches[3] === '1' ? TraceFlags.SAMPLED : TraceFlags.NONE,
+      isRemote: true,
+    };
+
```

Comment:
@kintel shoud we still validate the `traceId` and `spanId` have the correct length and are not zero. Even if the spec of the header dictates certain values the header can be manually populated as explained here: https://cloud.google.com/trace/docs/setup#force-trace

In that page there are no indications as to how many bits should the spanId have, nor the fact that both values must be different than 0.

What would happen in Opentelemetry if an invalid `spanContext` is set?

Reply:
Yes, we should manage invalid (zero) values according to the W3C spec: https://www.w3.org/TR/trace-context/#traceparent-header

The setup#force-trace page isn't great. A better reference is the API docs as they're auto-generated from the API implementation: https://cloud.google.com/trace/docs/reference/v2/rpc/google.devtools.cloudtrace.v2#google.devtools.cloudtrace.v2.Span

-> Trace ID = unsigned 128-bit non-zero value
-> Span ID = unsigned 64-bit non-zero value

Reply:
It's probably not the job of this propagator to auto-generate IDs though; propagators chan be chained and if there's a valid trace context elsewhere, we should let that propagator win.

Reply:
I think the correct solution would be to reject zero trace ID or span ID (return undefined). At least that's a clean first cut at this.

Reply:
> What would happen in Opentelemetry if an invalid spanContext is set?

OpenTelemetry only has [this](https://github.com/open-telemetry/opentelemetry-js/blob/e9b2cf9aeb1daf5ffbab800681bfe1cafc636576/packages/opentelemetry-core/src/trace/spancontext-utils.ts#L31) guard. I think it is job of propagator to provide valid `SpanContext`.

At-least we should validate `trace_id !== '00000000000000000000000000000000' && span_id !== '0000000000000000'` condition.

Reply:
Yeah, it's probably best to guard against zero IDs here

Reply:
done.